components/script/dom/node.rs

```diff
@@ -245,6 +243,7 @@ impl Node {
         let parent_in_doc = self.is_in_doc();
         for node in new_child.traverse_preorder() {
             node.set_flag(IS_IN_DOC, parent_in_doc);
+            debug_assert!(!node.get_flag(HAS_DIRTY_DESCENDANTS));
```

Comment:
Similarly, a comment here saying that we don't expect this flag to be added when we're out of the document, and how does styling and layout arrive to the element once it's attached would be helpful.